Patterns restructure (avoid Match/path mode terminology outside Reference section) (#1526)
neo4j/docs-cypher: modules/ROOT/pages/patterns/repeatable-relationship-and-node-matching.adoc

:description: Information about how to match patterns allowing re-traversals of both nodes and relationship using `REPEATABLE ELEMENTS`.
= Repeatable relationship and node matching
:table-caption!:
:page-role: new-neo4j-2025.06 cypher-25-only

By default, Cypher does not allow relationships to be re-traversed in the same graph pattern match.
This constraint can be bypassed using the `REPEATABLE ELEMENTS` keyword, which ensures there are no restrictions on how many times a node or relationship can occur for a given `MATCH` result.

The `REPEATABLE ELEMENTS` keyword must be specified after the `MATCH` clause.

.Query using `REPEATABLE ELEMENTS`
[source, cypher]
----
MATCH REPEATABLE ELEMENTS p = (start)--{,2}(end)
RETURN p
----

[NOTE]
`REPEATABLE ELEMENTS` is a xref:appendix/gql-conformance/index.adoc[GQL conformant] match mode.
Match modes determine whether relationships can appear more than once in a graph pattern match.

> For more information, see xref:patterns/reference.adoc#match-modes[Syntax and semantics -> Match modes]

Reviewer: ```suggestion
For more information, see xref:patterns/reference.adoc#match-modes[Syntax and semantics -> Match modes].
```

[[example-graph]]
== Example graph

This page will use the scenario of the link:https://en.wikipedia.org/wiki/Seven_Bridges_of_K%C3%B6nigsberg[Seven bridges of Königsberg], a problem studied by link:https://en.wikipedia.org/wiki/Leonhard_Euler[Leonhard Euler] in 1736 to determine if one could cross all seven bridges of Königsberg exactly once in a single walk.

image::match-modes-graph.svg[Graph showing the seven bridges of Königsberg, width=600, role=popup]

To recreate the graph, run the following query against an empty Neo4j database:

[source, cypher, role=test-setup]
----
CREATE (kneiphof:Location {name: "Kneiphof"}),
       (northBank:Location {name: "North Bank"}),
       (southBank:Location {name: "South Bank"}),
       (lomse:Location {name: "Lomse"}),

       (kneiphof)-[:BRIDGE {id: 1}]->(northBank),
       (kneiphof)-[:BRIDGE {id: 6}]->(southBank),
       (kneiphof)-[:BRIDGE {id: 7}]->(lomse),
       (northBank)-[:BRIDGE {id: 5}]->(kneiphof),
       (northBank)-[:BRIDGE {id: 2}]->(lomse),
       (southBank)-[:BRIDGE {id: 4}]->(kneiphof),
       (southBank)-[:BRIDGE {id: 3}]->(lomse)
----

[[repeatable-elements-and-konigsberg-bridges-problem]]

> == `REPEATABLE ELEMENTS` and the Köningsberg bridges problem

Reviewer 2: ```suggestion
== `REPEATABLE ELEMENTS` and the Königsberg bridges problem
```

The section on xref:patterns/unique-relationship-matching.adoc#default-traversal-rules-and-konigsberg-bridges-problem[Cypher's default traversal rules and the Königsberg bridges problem] showed the impossibility of crossing each of the seven bridges in Königsberg exactly once in a single walk.

`REPEATABLE ELEMENTS` does not solve this impossibility.

> However, the ability to re-traverse relationships does allow Cypher to return paths when testing Euler’s hypothesis.

Reviewer: ```suggestion
However, the ability to re-traverse relationships does allow Cypher to return paths when testing Euler's hypothesis.
```

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| samplePathLocations | samplePathBridges

| ["Kneiphof", "North Bank", "Kneiphof", "North Bank", "Kneiphof", "North Bank", "Kneiphof", "North Bank"]
| [1, 5, 1, 5, 1, 5, 1]

2+d|Rows: 1
|===

While no paths with a length of `7` can be returned using Cypher's default `MATCH` behavior, the query now returns a `samplePath` that shows the same two bridges (`1` and `5`) being repeatedly re-traversed until the maximum path length is reached.

The next example matches paths starting and ending at the same `Location` (`Kneiphof`), where each bridge is crossed at least once, using `REPEATABLE ELEMENTS`:

.Find a path traversing all bridges at least once before returning to the starting `Location` using `REPEATABLE ELEMENTS`
[source, cypher]
----
MATCH REPEATABLE ELEMENTS p = (start:Location {name: 'Kneiphof'})-[:BRIDGE]-{,9}(start) // <1>
WHERE all(bridge IN range(1,7) WHERE bridge IN [r IN relationships(p) | r.id]) // <2>
RETURN [n IN nodes(p) | n.name] AS visitedLocations,
       [r IN relationships(p) | r.id] AS crossedBridges
ORDER BY length(p), crossedBridges
LIMIT 1
----

> <1> Paths with a length of less than 9 relationships return no results because they do not allow enough moves to traverse all seven bridges at least once while forming a cycle back to Kneiphof.

Reviewer: ```suggestion
<1> Paths with a length of less than 9 relationships return no results because they do not allow enough moves to traverse all seven bridges at least once while forming a cycle back to `Kneiphof`.
```
